src/github.com/matrix-org/dendrite/common/config/appservice.go

```diff
@@ -157,12 +170,41 @@ func appendExclusiveNamespaceRegexs(
 
 // checkErrors checks for any configuration errors amongst the loaded
 // application services according to the application service spec.
+// nolint: gocyclo
 func checkErrors(config *Dendrite) (err error) {
 	var idMap = make(map[string]bool)
 	var tokenMap = make(map[string]bool)
 
+	// Compile regexp object for checking groupIDs
+	groupIDRegexp := regexp.MustCompile(`\+.*:.*`)
+
 	// Check each application service for any config errors
 	for _, appservice := range config.Derived.ApplicationServices {
+		// Check that namespace(s) are valid regex
+		for key, namespaceSlice := range appservice.NamespaceMap {
+			for _, namespace := range namespaceSlice {
+				if !IsValidRegex(namespace.Regex) {
+					return configErrors([]string{fmt.Sprintf(
+						"Invalid regex string for Application Service %s", appservice.ID,
+					)})
+				}
+
+				// Check if GroupID for the users namespace is in the correct format
+				if key == "users" && namespace.GroupID != "" {
```

Comment:
would looping over `appservice.NamespaceMap["users"]` again and removing the if statement reduce cyclomatic complexity ?

I know gocyclo is a pain. Still cyclo of over 12 should be avoided.

Reply:
Seems to make it even worse, 15 > 14.

Reply:
Usually the thing to do is to move logic out into a separate function, e.g. pull out the inner loop contents into a `ValidatAppserviceConfig` func

Reply:
The `checkErrors` function essentially was `ValidateAppserviceConfig`. I've moved the namespace checks into a separate function and that's made gocyclo happy.